features/telemetry/protocols/netflow/parser/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/parser/RecordEnricher.java

```diff
@@ -0,0 +1,193 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.telemetry.protocols.netflow.parser;
+
+import java.net.InetAddress;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import org.opennms.netmgt.dnsresolver.api.DnsResolver;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.Value;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.BooleanValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.DateTimeValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.FloatValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.IPv4AddressValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.IPv6AddressValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.ListValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.MacAddressValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.NullValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.OctetArrayValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.SignedValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.StringValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.UndeclaredValue;
+import org.opennms.netmgt.telemetry.protocols.netflow.parser.ie.values.UnsignedValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Sets;
+
+public class RecordEnricher {
+    private static final Logger LOG = LoggerFactory.getLogger(RecordEnricher.class);
+
+    private final DnsResolver dnsResolver;
+
+    public RecordEnricher(DnsResolver dnsResolver) {
+        this.dnsResolver = Objects.requireNonNull(dnsResolver);
+    }
+
+    public CompletableFuture<RecordEnrichment> enrich(Iterable<Value<?>> record) {
+        final IpAddressCapturingVisitor ipAddressCapturingVisitor = new IpAddressCapturingVisitor();
+        for (final Value<?> value : record) {
+            value.visit(ipAddressCapturingVisitor);
+        }
+        final Set<InetAddress> addressesToReverseLookup = ipAddressCapturingVisitor.getAddresses();
+        final Map<InetAddress, String> hostnamesByAddress = new HashMap<>(addressesToReverseLookup.size());
```

Comment:
Not sure I follow here.

Reply:
I'm trying to avoid duplicate lookups (even if they are cached) and try to get rid of the synchronize point.
It's just a vague proposal
```
        final Map<InetAddress, CompletableFuture<String>> hostnamesByAddress = new HashMap<>(addressesToReverseLookup.size());
        final CompletableFuture reverseLookupFutures[] = addressesToReverseLookup.stream()
                .map(addr -> {
                    return hostnamesByAddress.computeIfAbsent(addr, addr -> {
                        return dnsResolver.reverseLookup(addr).whenComplete((hostname, ex) -> {...});
                    });
                }).toArray(CompletableFuture[]::new);
```

Access the the hostnames can just be done by `hostnamesByAddress.get(ip).get()` or in a transformation after the overall future has completed. (Maybe this can also defere the futures resolution to the time the packet is serialized).

As said, this is just a vague idea. Maybe I have to come up with something that compiles...

Reply:
We start with a set of unique set IP addresses. What do we need to dedup?

Maybe we should do this one layer up instead? We could perform a single enrichment per packet instead of one per record.

Reply:
We start with a set of unique set IP addresses. What do we need to dedup?

Maybe we should do this one layer up instead? We could perform a single enrichment per packet instead of one per record.

Reply:
Oh. I've missed the `Set` completely. You're totally right...

Reply:
Oh. I've missed the `Set` completely. You're totally right...